File: src/main/java/id/walt/ssikitexamples/CustomDataAndPolicy.java
package id.walt.ssikitexamples;

import com.fasterxml.jackson.annotation.JsonIgnore;
import id.walt.auditor.Auditor;
import id.walt.auditor.SimpleVerificationPolicy;
import id.walt.auditor.VerificationPolicyResult;
import id.walt.auditor.policies.JsonSchemaPolicy;
import id.walt.auditor.policies.SignaturePolicy;
import id.walt.credentials.w3c.PresentableCredential;
import id.walt.credentials.w3c.VerifiableCredential;
import id.walt.credentials.w3c.W3CIssuer;
import id.walt.credentials.w3c.builder.W3CCredentialBuilder;
import id.walt.custodian.Custodian;
import id.walt.servicematrix.ServiceMatrix;
import id.walt.signatory.*;
import kotlin.Unit;
import org.jetbrains.annotations.NotNull;

import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static java.util.Optional.ofNullable;

public class CustomDataAndPolicy {

    public static void main(String[] args) {
        new CustomDataAndPolicy().run();
    }

    public void run() {
        // Load walt.id SSI Kit services from "$workingDirectory/service-matrix.properties"
        new ServiceMatrix("service-matrix.properties");

        // Define used services
        var signatory = Signatory.Companion.getService();
        var custodian = Custodian.Companion.getService();

        // Create VCs to verify:
        var idIter = MockedIdDatabase.INSTANCE.getMockedIds().values().iterator();
        var holder = idIter.next();
        var issuer = idIter.next();

        // Issue VC in JSON-LD and JWT format
        var vcJsonLd = signatory.issue("VerifiableId", createProofConfig(issuer.getDid(), holder.getDid(), ProofType.LD_PROOF, holder.getPersonalIdentifier()), new IdDataCustomProvider(), null, false);
        System.out.println("\n------------------------------- VC in JSON-LD format -------------------------------");
        System.out.println(vcJsonLd);

        var vcJwt = signatory.issue("VerifiableId", createProofConfig(issuer.getDid(), holder.getDid(), ProofType.JWT, holder.getPersonalIdentifier()), new IdDataCustomProvider(), null, false);
        System.out.println("\n------------------------------- VC in JWT format -------------------------------");
        System.out.println(vcJwt);

        // Verify VPs, using Signature, JsonSchema and a custom policy
        var resVcJsonLd = Auditor.Companion.getService().verify(vcJsonLd, List.of(new SignaturePolicy(), new JsonSchemaPolicy(), new CustomPolicy()));
        var resVcJwt = Auditor.Companion.getService().verify(vcJwt, List.of(new SignaturePolicy(), new JsonSchemaPolicy(), new CustomPolicy()));

        System.out.println("JSON verification result: " + resVcJsonLd.getResult());
        System.out.println("JWT verification result: " + resVcJwt.getResult());

        // Present VC in JSON-LD and JWT format
        var vpJsonLd = custodian.createPresentation(List.of(new PresentableCredential(VerifiableCredential.Companion.fromString(vcJsonLd), null, false)), holder.getDid(), null, null, null, null);
        System.out.println("------------------------------- VP in JSON-LD format -------------------------------");
        System.out.println(vpJsonLd);

        var vpJwt = custodian.createPresentation(List.of(new PresentableCredential(VerifiableCredential.Companion.fromString(vcJwt), null, false)), holder.getDid(), null, null, null, null);
        System.out.println("\n------------------------------- VP in JWT format -------------------------------");
        System.out.println(vpJwt);

        // Verify VPs, using Signature, JsonSchema and a custom policy
        var resJsonLd = Auditor.Companion.getService().verify(vpJsonLd, List.of(new SignaturePolicy(), new JsonSchemaPolicy(), new CustomPolicy()));
        var resJwt = Auditor.Companion.getService().verify(vpJwt, List.of(new SignaturePolicy(), new JsonSchemaPolicy(), new CustomPolicy()));

        System.out.println("JSON verification result: " + resJsonLd.getResult());
        System.out.println("JWT verification result: " + resJwt.getResult());
    }

    public ProofConfig createProofConfig(String issuerDid, String subjectDid, ProofType proofType, String dataProviderIdentifier) {
        return new ProofConfig(issuerDid, subjectDid, null, null, proofType, null, null,
                null, null, null, null, null, dataProviderIdentifier, null, null, Ecosystem.DEFAULT,
                null, "", "", null);
    }
}

class CustomPolicy extends SimpleVerificationPolicy {

    private String description = "A custom verification policy";

    @NotNull
    @Override
    public String getDescription() {
        return description;
    }

    @NotNull
    @Override
    protected VerificationPolicyResult doVerify(@NotNull VerifiableCredential vc) {
        IdData idData = null;
        var result = false;
        if (vc.getType().get(vc.getType().size() - 1).equals("VerifiableId")) {
            if (vc.getCredentialSubject() != null) {
                idData = MockedIdDatabase.INSTANCE.get((String) vc.getCredentialSubject().getProperties().get("personalIdentifier"));
            }
            if (idData != null) {
                result = idData.getFamilyName().equals(vc.getCredentialSubject().getProperties().get("familyName"))
                        && idData.getFirstName().equals(vc.getCredentialSubject().getProperties().get("firstName"));
            }
        } else if (vc.getType().get(vc.getType().size() - 1).equals("VerifiablePresentation")) {
            // This custom policy does not verify the VerifiablePresentation
            result = true;
        }
        return result ? VerificationPolicyResult.Companion.success() : VerificationPolicyResult.Companion.failure(new Exception(""));
    }
}

class IdDataCustomProvider implements SignatoryDataProvider {

    @JsonIgnore
    private DateTimeFormatter dateFormat = DateTimeFormatter.ISO_INSTANT;

    @NotNull
    @Override
    public W3CCredentialBuilder populate(@NotNull W3CCredentialBuilder credentialBuilder, @NotNull ProofConfig proofConfig) {
        if (credentialBuilder.getType().get(credentialBuilder.getType().size() - 1).equals("VerifiableId")) {

            var idData = ofNullable(proofConfig.getDataProviderIdentifier())
                    .map(MockedIdDatabase.INSTANCE::get)
                    .orElseThrow(() -> new RuntimeException("No ID data found for the given data-provider identifier"));

            return updateCustomVerifiableCredential(credentialBuilder, proofConfig, idData);
        } else {
            throw new IllegalArgumentException("Only VerifiableId is supported by this data provider");
        }
    }

    private W3CCredentialBuilder updateCustomVerifiableCredential(W3CCredentialBuilder credentialBuilder, ProofConfig proofConfig, IdData idData) {
        return credentialBuilder
                .setId("identity#verifiableID#" + UUID.randomUUID())
                .setIssuer(new W3CIssuer(proofConfig.getIssuerDid()))
                .setProperty("evidence", List.of(Map.of(
                        "verifier", proofConfig.getIssuerDid())
                ))
                .buildSubject(subjectBuilder -> {
                    subjectBuilder.setId(idData.getDid());
                    subjectBuilder.setProperty("familyName", idData.getFamilyName());
                    subjectBuilder.setProperty("firstName", idData.getFirstName());
                    subjectBuilder.setProperty("dateOfBirth", idData.getDateOfBirth());
                    subjectBuilder.setProperty("personalIdentifier", idData.getPersonalIdentifier());
                    subjectBuilder.setProperty("nameAndFamilyNameAtBirth", idData.getNameAndFamilyNameAtBirth());
                    subjectBuilder.setProperty("placeOfBirth", idData.getPlaceOfBirth());
                    subjectBuilder.setProperty("currentAddress", Collections.singletonList(idData.getCurrentAddress()));
                    subjectBuilder.setProperty("gender", idData.getGender());

                    return Unit.INSTANCE;
                }).setIssuanceDate(
                        proofConfig.getIssueDate() != null ? Instant.parse(dateFormat.format(proofConfig.getIssueDate())) : Instant.now()
                )
                .setExpirationDate(
                        proofConfig.getExpirationDate() != null ? Instant.parse(dateFormat.format(proofConfig.getExpirationDate())) : Instant.now()
                )
                .setValidFrom(
                        proofConfig.getIssueDate() != null ? Instant.parse(dateFormat.format(proofConfig.getExpirationDate())) : Instant.now()
                );
    }
}

File: src/main/kotlin/id/walt/ssikitexamples/CustomDataAndPolicy.kt
package id.walt.ssikitexamples

import com.beust.klaxon.Json
import id.walt.auditor.Auditor
import id.walt.auditor.SimpleVerificationPolicy
import id.walt.auditor.VerificationPolicyResult
import id.walt.auditor.policies.JsonSchemaPolicy
import id.walt.auditor.policies.SignaturePolicy
import id.walt.credentials.w3c.PresentableCredential
import id.walt.credentials.w3c.VerifiableCredential
import id.walt.credentials.w3c.builder.W3CCredentialBuilder
import id.walt.custodian.Custodian
import id.walt.servicematrix.ServiceMatrix
import id.walt.signatory.ProofConfig
import id.walt.signatory.ProofType
import id.walt.signatory.Signatory
import id.walt.signatory.SignatoryDataProvider
import java.time.Instant
import java.time.format.DateTimeFormatter
import java.util.*

val signatory = Signatory.getService()
val custodian = Custodian.getService()

fun main(){
    customDataAndPolicy()
}

fun customDataAndPolicy() {
    // Load walt.id SSI-Kit services from "$workingDirectory/service-matrix.properties"
    ServiceMatrix("service-matrix.properties")

    // Create VCs to verify:
    val idIter = MockedIdDatabase.mockedIds.values.iterator()
    val holder = idIter.next()
    val issuer = idIter.next()

    // Issue VC in JSON-LD and JWT format
    val vcJsonLd = signatory.issue(
        "VerifiableId",
        ProofConfig(issuerDid = issuer.did, subjectDid = holder.did, proofType = ProofType.LD_PROOF, dataProviderIdentifier = holder.personalIdentifier),
        dataProvider = CustomIdDataProvider()
    )
    println("\n------------------------------- VC in JSON_LD format -------------------------------")
    println(vcJsonLd)
    val vcJwt = signatory.issue(
        "VerifiableId", ProofConfig(
            issuerDid = issuer.did,
            subjectDid = holder.did,
            proofType = ProofType.JWT,
            dataProviderIdentifier = holder.personalIdentifier
        ), dataProvider = CustomIdDataProvider()
    )
    println("\n------------------------------- VC in JWT format -------------------------------")
    println(vcJwt)

    // Verify VPs, using Signature, JsonSchema and a custom policy
    val resVcJson = Auditor.getService().verify(vcJsonLd, listOf(SignaturePolicy(), JsonSchemaPolicy(), MyCustomPolicy()))
    val resVcJwt = Auditor.getService().verify(vcJwt, listOf(SignaturePolicy(), JsonSchemaPolicy(), MyCustomPolicy()))

    println("JSON verification result: ${resVcJson.result}")
    println("JWT verification result: ${resVcJwt.result}")

    // Present VC in JSON-LD and JWT format
    val vpJson = custodian.createPresentation(
        listOf(PresentableCredential(VerifiableCredential.fromString(vcJsonLd))), holder.did, null, null, null, null
    )
    println("------------------------------- VP in JSON_LD format -------------------------------")
    println(vpJson)
    val vpJwt = custodian.createPresentation(
        listOf(PresentableCredential(VerifiableCredential.fromString(vcJwt))), holder.did, null, null, null, null
    )
    println("\n------------------------------- VP in JWT format -------------------------------")
    println(vpJwt)

    // Verify VPs, using Signature, JsonSchema and a custom policy
    val resJson = Auditor.getService().verify(vpJson, listOf(SignaturePolicy(), JsonSchemaPolicy(), MyCustomPolicy()))
    val resJwt = Auditor.getService().verify(vpJwt, listOf(SignaturePolicy(), JsonSchemaPolicy(), MyCustomPolicy()))

    println("JSON verification result: ${resJson.result}")
    println("JWT verification result: ${resJwt.result}")
}


// Custom Policy
class MyCustomPolicy : SimpleVerificationPolicy() {
    override val description: String
        get() = "A custom verification policy"

    override fun doVerify(vc: VerifiableCredential): VerificationPolicyResult {
        return when (vc.type.last()) {
            "VerifiableId" -> MockedIdDatabase.get(vc.credentialSubject!!.properties["personalIdentifier"] as String)
                ?.takeIf {
                    it.familyName == vc.credentialSubject!!.properties["familyName"] && it.firstName == vc.credentialSubject!!.properties["firstName"]
                }?.let { VerificationPolicyResult.success() } ?: VerificationPolicyResult.failure(Exception(""))

            "VerifiablePresentation" -> {
                // This custom policy does not verify the VerifiablePresentation
                VerificationPolicyResult.success()
            }
            else -> VerificationPolicyResult.failure(IllegalArgumentException(""))
        }
    }
}

// Custom Data Provider
class CustomIdDataProvider : SignatoryDataProvider {

    @Json(ignored = true)
    val dateFormat = DateTimeFormatter.ISO_INSTANT!!

    override fun populate(credentialBuilder: W3CCredentialBuilder, proofConfig: ProofConfig): W3CCredentialBuilder {
        return when(credentialBuilder.type.last()) {
            "VerifiableId" -> {
                credentialBuilder.setId("identity#verifiableID#${UUID.randomUUID()}")
                    .setIssuer(proofConfig.issuerDid)
                    .setIssuanceDate(proofConfig.issueDate?.let { Instant.parse(dateFormat.format(it)) } ?: Instant.now())
                    .setExpirationDate(proofConfig.expirationDate?.let { Instant.parse(dateFormat.format(it)) } ?: Instant.now())
                    .setValidFrom(proofConfig.issueDate?.let { Instant.parse(dateFormat.format(it)) } ?: Instant.now())
                    .setProperty("evidence", buildList {
                        add(mapOf("verifier" to proofConfig.issuerDid))
                    }).buildSubject {
                        // get ID data for the given subject
                        val idData = MockedIdDatabase.get(proofConfig.dataProviderIdentifier!!)
                            ?: throw Exception("No ID data found for the given data-povider identifier")
                        setProperty("familyName", idData.familyName)
                        setProperty("firstName", idData.firstName)
                        setProperty("dateOfBirth", idData.dateOfBirth)
                        setProperty("personalIdentifier", idData.personalIdentifier)
                        setProperty("nameAndFamilyNameAtBirth", idData.nameAndFamilyNameAtBirth)
                        setProperty("placeOfBirth", idData.placeOfBirth)
                        setProperty("currentAddress", listOf(idData.currentAddress))
                        setProperty("gender", idData.gender)
                    }
            }

            else -> throw IllegalArgumentException("Only VerifiableId is supported by this data provider");
        }
    }
}